// knip/dist/typescript/visitors/dynamic-imports/requireCall.js
import ts from 'typescript';
import { IMPORT_FLAGS, IMPORT_STAR } from '../../../constants.js';
import { findAncestor, findDescendants, isModuleExportsAccess, isRequireCall, isTopLevel } from '../../ast-helpers.js';
import { isNotJS } from '../helpers.js';
import { importVisitor as visit } from '../index.js';
export default visit(() => true, node => {
    if (isRequireCall(node)) {
        if (ts.isStringLiteralLike(node.arguments[0])) {
            const specifier = node.arguments[0].text;
            const modifiers = isNotJS(node.getSourceFile()) ? IMPORT_FLAGS.BRIDGE : IMPORT_FLAGS.NONE;
            if (specifier) {
                const propertyAccessExpression = findAncestor(node, _node => {
                    if (ts.isExpressionStatement(_node) || ts.isCallExpression(_node))
                        return 'STOP';
                    return ts.isPropertyAccessExpression(_node);
                });
                if (propertyAccessExpression) {
                    const identifier = String(propertyAccessExpression.name.escapedText);
                    return {
                        identifier,
                        specifier,
                        pos: propertyAccessExpression.name.getStart(),
                        modifiers,
                        alias: undefined,
                        namespace: undefined,
                        symbol: undefined,
                    };
                }
                const variableDeclaration = node.parent;
                if (ts.isVariableDeclaration(variableDeclaration) &&
                    ts.isVariableDeclarationList(variableDeclaration.parent)) {
                    const isTLA = isTopLevel(variableDeclaration.parent);
                    if (ts.isIdentifier(variableDeclaration.name)) {
                        const alias = String(variableDeclaration.name.escapedText);
                        return {
                            identifier: 'default',
                            alias,
                            symbol: isTLA ? variableDeclaration.symbol : undefined,
                            specifier,
                            pos: variableDeclaration.name.getStart(),
                            namespace: undefined,
                            modifiers,
                        };
                    }
                    const bindings = findDescendants(variableDeclaration, ts.isBindingElement);
                    if (bindings.length > 0) {
                        return bindings.map(element => {
                            const identifier = (element.propertyName ?? element.name).getText();
                            const alias = element.propertyName ? element.name.getText() : undefined;
                            const symbol = isTLA ? element.symbol : undefined;
                            return {
                                identifier,
                                specifier,
                                alias,
                                symbol,
                                pos: element.name.getStart(),
                                modifiers,
                                namespace: undefined,
                            };
                        });
                    }
                    return {
                        identifier: 'default',
                        specifier,
                        pos: node.arguments[0].pos,
                        modifiers,
                        alias: undefined,
                        namespace: undefined,
                        symbol: undefined,
                    };
                }
                if (ts.isBinaryExpression(node.parent) &&
                    ts.isPropertyAccessExpression(node.parent.left) &&
                    isModuleExportsAccess(node.parent.left)) {
                    return {
                        identifier: IMPORT_STAR,
                        specifier,
                        pos: node.arguments[0].pos,
                        modifiers: IMPORT_FLAGS.RE_EXPORT,
                        alias: undefined,
                        namespace: undefined,
                        symbol: undefined,
                    };
                }
                if (ts.isCallExpression(node.parent)) {
                    return {
                        identifier: 'default',
                        specifier,
                        pos: node.getEnd(),
                        modifiers,
                        alias: undefined,
                        namespace: undefined,
                        symbol: undefined,
                    };
                }
                return {
                    identifier: 'default',
                    specifier,
                    pos: node.getStart(),
                    modifiers,
                    alias: undefined,
                    namespace: undefined,
                    symbol: undefined,
                };
            }
        }
    }
});
