// knip/dist/typescript/visitors/imports/reExportDeclaration.js
import ts from 'typescript';
import { IMPORT_FLAGS, IMPORT_STAR } from '../../../constants.js';
import { importVisitor as visit } from '../index.js';
export default visit(() => true, node => {
    if (ts.isExportDeclaration(node)) {
        if (node.moduleSpecifier && ts.isStringLiteralLike(node.moduleSpecifier)) {
            if (!node.exportClause) {
                return {
                    identifier: IMPORT_STAR,
                    specifier: node.moduleSpecifier.text,
                    pos: node.moduleSpecifier.getStart() - 7,
                    modifiers: IMPORT_FLAGS.RE_EXPORT,
                    alias: undefined,
                    namespace: undefined,
                    symbol: undefined,
                };
            }
            if (node.exportClause.kind === ts.SyntaxKind.NamespaceExport) {
                return {
                    identifier: IMPORT_STAR,
                    namespace: String(node.exportClause.name.text),
                    specifier: node.moduleSpecifier.text,
                    pos: node.exportClause.name.getStart(),
                    modifiers: IMPORT_FLAGS.RE_EXPORT,
                    alias: undefined,
                    symbol: undefined,
                };
            }
            const specifier = node.moduleSpecifier;
            return node.exportClause.elements.map(element => {
                if (element.propertyName && element.name) {
                    return {
                        identifier: String(element.propertyName.text),
                        alias: String(element.name.text),
                        specifier: specifier.text,
                        pos: element.propertyName.getStart(),
                        modifiers: IMPORT_FLAGS.RE_EXPORT,
                        namespace: undefined,
                        symbol: undefined,
                    };
                }
                return {
                    identifier: (element.propertyName ?? element.name).getText(),
                    specifier: specifier.text,
                    pos: element.name.getStart(),
                    modifiers: IMPORT_FLAGS.RE_EXPORT,
                    alias: undefined,
                    namespace: undefined,
                    symbol: undefined,
                };
            });
        }
    }
});
