// @vue-macros/nuxt/dist/index.js
import { defineNuxtModule, useNuxt } from "@nuxt/kit";
import { REGEX_SETUP_SFC } from "@vue-macros/common";
import { resolveOptions } from "vue-macros";
import VueMacros from "vue-macros/vite";

//#region src/index.ts
const module = defineNuxtModule({
	meta: {
		name: "vue-macros",
		configKey: "macros"
	},
	defaults: {},
	async setup(options) {
		const nuxt = useNuxt();
		const resolvedOptions = await resolveOptions(options);
		nuxt.options.typescript.tsConfig ||= {};
		const vueCompilerOptions = nuxt.options.typescript.tsConfig.vueCompilerOptions ||= {};
		vueCompilerOptions.plugins ||= [];
		vueCompilerOptions.plugins.push("@vue-macros/nuxt/volar");
		nuxt.hook("vite:configResolved", async (config, { isClient }) => {
			function findPluginAndRemove(name) {
				const idx$1 = config.plugins.findIndex((plugin$1) => plugin$1 && "name" in plugin$1 && plugin$1.name === name);
				if (idx$1 === -1) return [null, idx$1];
				const plugin = config.plugins[idx$1];
				config.plugins.splice(idx$1, 1);
				return [plugin, idx$1];
			}
			config.plugins ||= [];
			const [vue, idx] = findPluginAndRemove("vite:vue");
			const [vueJsx] = findPluginAndRemove("vite:vue-jsx");
			const vueMacrosPlugins = await VueMacros({
				...resolvedOptions,
				plugins: {
					vue,
					vueJsx
				},
				nuxtContext: { isClient }
			});
			if (idx === -1) config.plugins.push(...vueMacrosPlugins);
			else config.plugins.splice(idx, 0, ...vueMacrosPlugins);
		});
		nuxt.hook("prepare:types", (opts) => {
			opts.references.push({ types: "vue-macros/macros-global" });
		});
		nuxt.hook("devtools:customTabs", (tabs) => {
			tabs.push({
				name: "vue-macros",
				title: "Vue Macros",
				icon: `https://raw.githubusercontent.com/vue-macros/vue-macros/main/docs/public/favicon.svg`,
				view: {
					type: "iframe",
					src: "/__vue-macros"
				}
			});
		});
		const viteVue = nuxt.options.vite.vue ||= {};
		if (resolvedOptions.setupSFC) {
			viteVue.include ||= [/\.vue$/];
			if (!Array.isArray(viteVue.include)) viteVue.include = [viteVue.include];
			viteVue.include.push(REGEX_SETUP_SFC);
			nuxt.hook("components:extend", (components) => {
				for (const component of components) {
					component.pascalName = component.pascalName.replace(/Setup$/, "");
					component.kebabName = component.kebabName.replace(/-setup$/, "");
				}
			});
			nuxt.hook("pages:extend", (pages) => {
				for (const page of pages) {
					if (!page.file || !REGEX_SETUP_SFC.test(page.file)) continue;
					if (page.name) page.name = page.name.replace(/\.setup$/, "");
					if (page.path) page.path = page.path.replace(/\/index\.setup$/, "/").replace(/\.setup$/, "");
				}
			});
		}
		nuxt.hook("app:resolve", (app) => {
			app.layouts = Object.fromEntries(Object.entries(app.layouts).map(([key, value]) => {
				if (key.endsWith("-setup")) {
					key = key.replace(/-setup$/, "");
					value.name = value.name.replace(/-setup$/, "");
				}
				return [key, value];
			}));
		});
	}
});
var src_default = module;

//#endregion
export { src_default as default };
